src/dali/plugin/input/rest/binance_com.py

```diff
@@ -0,0 +1,774 @@
+        #    {
+        #      "code": "000000",
+        #      "message": "success",
+        #      "data": [
+        #        {
+        #          "orderNo": "25ced37075c1470ba8939d0df2316e23",
+        #          "fiatCurrency": "EUR",
+        #          "indicatedAmount": "15.00",
+        #          "amount": "15.00",
+        #          "totalFee": "0.00",
+        #          "method": "card",
+        #          "status": "Failed",
+        #          "createTime": 1627501026000,
+        #          "updateTime": 1627501027000
+        #        }
+        #      ],
+        #      "total": 1,
+        #      "success": True
+        #    }
+        if _DATA in fiat_deposits:
+            for deposit in fiat_deposits[_DATA]:
+                if deposit[_STATUS] == "Completed":
+                    self._process_deposit(deposit, in_transactions)
+
+    def _process_gains(self, in_transactions: List[InTransaction]) -> None:
+
+        ### Regular Dividends from Staking (including Eth staking) and Savings (Lending)
+
+        # We need milliseconds for Binance
+        current_start = self.start_time_ms
+        now_time = int(datetime.datetime.now().timestamp()) * 1000
+
+        # We will pull in 30 day periods. This allows for 16 assets with daily dividends.
+        current_end = current_start + 2592000000
+
+        while current_start < now_time:
+
+            # CCXT doesn't have a standard way to pull income, we must use the underlying API call
+            dividends = self.client.sapiGetAssetAssetDividend(params=({_STARTTIME: current_start, _ENDTIME: current_end, _LIMIT: 500}))
+            # {
+            #     "rows":[
+            #         {
+            #             "id":1637366104,
+            #             "amount":"10.00000000",
+            #             "asset":"BHFT",
+            #             "divTime":1563189166000,
+            #             "enInfo":"BHFT distribution",
+            #             "tranId":2968885920
+            #         },
+            #         {
+            #             "id":1631750237,
+            #             "amount":"10.00000000",
+            #             "asset":"BHFT",
+            #             "divTime":1563189165000,
+            #             "enInfo":"BHFT distribution",
+            #             "tranId":2968885920
+            #         }
+            #     ],
+            #     "total":2
+            # }
+            for dividend in dividends[_ROWS]:
+                if dividend[_ENINFO] in _STAKING:
+                    self._process_gain(dividend, Keyword.STAKING, in_transactions)
+                elif dividend[_ENINFO] in _INTEREST:
+                    self._process_gain(dividend, Keyword.INTEREST, in_transactions)
+                else:
+                    self.__logger.error("WARNING: Unrecognized Dividend: %s. Please open an issue at %s", dividend[_ENINFO], self.ISSUES_URL)
+                    self._process_gain(dividend, Keyword.INCOME, in_transactions)

```

Comment:
I wasn't sure what to do here since there could potentially be hundreds of different products. Every airdrop is a different product for example. However, everything coming out of `sapiGetAssetAssetDividend` is some kind of income. So, we have two choices - 1) label it as income, throw an error, and let the user sort it out until we fix it or 2) stop processing and have them report an issue and we send out a PR with the fix, which could delay the user being able to do their taxes for a few days to a week.

It's a relatively small fix to make to add a new dividend. As long as someone is around to patch it up, it wouldn't take long.

Personally, I would like to see the report with mislabeled dividends so that I have something while I wait for the fix. That is why I added this code. However, thinking about it now, 2) might be better because it more or less forces the user to report it, and that benefits other users. 

What do you think?

Reply:
Got it, thanks for the explanation. My approach has been to throw errors and asking users to open issues, so that we always know what gets processed and how: so I would prefer option 2. This seems to work OK for Coinbase, but I'm not sure about Binance, especially since you mentioned there may be 100s of cases. Potentially we could do 2, see how it goes and if it's unmanageable we can always switch back to 1. However it's just a suggestion, because I don't have a Binance account and so I don't have a first hand pulse of the problem. This means you get to have the last word on how to approach this :-)